Add info about type hints rendering to the docstrings section (#101)
pyansys/dev-guide: doc/source/doc-style/docstrings.rst

----------

> Functions and class methods may have parameters in their signatures. All these

Reviewer: I still don't know how to distinguish between methods and functions.

> Functions and class methods may have parameters in their signatures. All these

Author: Indeed, the two are very similar! To be precise with the language:

- Function is the general term, it includes both *methods* and *free functions*:
- *methods* are functions which are "attached" to a class
- *free functions* in contrast do not belong to a class

The sentence above also mentions "class methods" -- which is again a subset of "methods"; they are attached to a class, but can be called from the class directly, as opposed to "instance methods", which need an _instance_ of a class.

In more relaxed / sloppy speech, "function" can refer to "free function" if it's clear from context.

So again, as a tree

- Functions
  - Free functions
  - Methods
    - Class methods
    - Instance methods
    - Static methods

In general, the [Python glossary](https://docs.python.org/3/glossary.html) is a good place to look up such terms. It's a bit annoying, but each programming language can have slightly different ways in which things are named.

Long story short: For the sentence above, I think it would be sufficient to say

``Functions can have parameters...``

To be pedantic, classes _also_ have signatures, but that is the same as the signature of their constructor (a special method).

    student_version : bool, optional
        Whether to enable the student version of AEDT. The default
        is ``False``.

The name of each parameter is followed by a space, a colon, a space, and then

> the data type. A parameter is optional if its keyword argument has a default shown
> in the function signature. For an optional parameter, the data type is followed by a

Reviewer: ```suggestion
the data type. A parameter is optional if its keyword argument displays a default
in the function signature. For an optional parameter, the data type is followed by a
```

Author: ```suggestion
the data type. A parameter is optional if a default is defined in its signature. 
For an optional parameter, the data type is followed by a
```

In this sentence, the "keyword argument" seems misplaced to me. Not all parameters are keyword arguments, see https://docs.python.org/3/glossary.html#term-argument

Author: @PipKat I'm not sure if we should mix general edits to this page together with the addition of the type hint section.

Note that the changes I've made to the rest of the document are just the removal of trailing whitespace.

If the changes are small enough I guess we can do it in one go, but I wouldn't want to delay this PR with other changes.

Reviewer: Feel free to go back to your original text on these lines. The number of edits are small--so I'd suggest making them.

If the definition of the function is updated, this
section must be updated too.

Type Hints
==========

By default, Sphinx renders `type hints <https://peps.python.org/pep-0484/>`_ as part
of the function signature. This can become difficult to read, because the signature
becomes very long.

> Instead, we can render the type hints as part of each parameter's description. For
> this, the ``sphinx.ext.autodoc.typehints``, ``sphinx.ext.napoleon``, and
> ``numpydoc`` extensions need to be combined as follows (in ``conf.py``):

Reviewer: ```suggestion
Instead, we can render the type hints as part of each parameter's description. To
accomplish this, the ``sphinx.ext.autodoc.typehints``, ``sphinx.ext.napoleon``, and
``numpydoc`` extensions must be combined in the ``conf.py`` file as follows:
```

.. code:: python

  extensions = [
      ...
      "sphinx.ext.autodoc.typehints",
      "sphinx.ext.napoleon",
      "numpydoc",
      ...
  ]
  autodoc_typehints = "description"

> Note that the order in which the extensions are included matters.
> 

Reviewer: ```suggestion
```

Reviewer 2: We might consider keeping this and having this within a `note` directive:

```suggestion
.. note::
   The order in which the extensions are included matters.

```

Reviewer: @akaszynski Keeping it is OK with me--I just changed the above to say that had to appear in the following order so viewed the note as redundant. However, it emphasizes the fact, which may be your intent.

Author: Ah, seems I messed this up by prematurely accepting the edits. I'll revert it tomorrow, when I'm in front of my usual PC.

@PipKat if we keep the note, would you lean towards keeping the edited version of the sentence above, or revert it to the original?

Reviewer: The note directive will definitely put more emphasis on this information. If calling it out is important, use the note!

Author: Ok, yeah I think we should use the note. In that case should we revert to the original 
```
Instead, we can render the type hints as part of each parameter's description. For
this, the ``sphinx.ext.autodoc.typehints``, ``sphinx.ext.napoleon``, and
``numpydoc`` extensions need to be combined as follows (in ``conf.py``):
```
or use 
```
Instead, we can render the type hints as part of each parameter's description. To
accomplish this, the ``sphinx.ext.autodoc.typehints``, ``sphinx.ext.napoleon``, and
``numpydoc`` extensions must be combined in the ``conf.py`` file in this order:
```

To me, the latter version now looks somewhat redundant, but wanted to hear your opinion on that.

Reviewer: I'd do the last--remove "in this order" from the suggested edit.
